File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_set = set()
        left = 0 
        longest = 0

        for right in range(len(s)):
          while s[right] in char_set:
            char_set.remove(s[left])
            left += 1

          char_set.add(s[right])
          current_window = right - left + 1 
          longest = max(longest, current_window)
         
        return longest 

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
       int longest = 0;
       int left = 0;
       Set<Character> seen = new HashSet<>();
       for (int right=0; right<s.length(); right++){
        char c = s.charAt(right);
        while(seen.contains(c)){
          seen.remove(s.charAt(left));
          left++;
        }
        seen.add(s.charAt(right));
        longest = Math.max(longest, right-left+1);
       } 
       return longest;
    }
}